library/test/src/formatters/json.rs

```diff
@@ -39,29 +35,35 @@ impl<T: Write> JsonFormatter<T> {
         stdout: Option<Cow<'_, str>>,
         extra: Option<&str>,
     ) -> io::Result<()> {
+        let mut out = Vec::new();
+
         // A doc test's name includes a filename which must be escaped for correct json.
-        self.write_message(&format!(
+        write!(&mut out,
             r#"{{ "type": "{}", "name": "{}", "event": "{}""#,
             ty,
             EscapedString(name),
             evt
-        ))?;
+        )?;
         if let Some(exec_time) = exec_time {
-            self.write_message(&format!(r#", "exec_time": {}"#, exec_time.0.as_secs_f64()))?;
+            write!(&mut out, r#", "exec_time": {}"#, exec_time.0.as_secs_f64())?;
         }
         if let Some(stdout) = stdout {
-            self.write_message(&format!(r#", "stdout": "{}""#, EscapedString(stdout)))?;
+            write!(&mut out, r#", "stdout": "{}""#, EscapedString(stdout))?;
         }
         if let Some(extra) = extra {
-            self.write_message(&format!(r#", {extra}"#))?;
+            write!(&mut out, r#", {extra}"#)?;
         }
-        self.writeln_message(" }")
+        write!(&mut out, "}}\n")?;
+        self.writeln_message(&out)
     }
 }
 
 impl<T: Write> OutputFormatter for JsonFormatter<T> {
     fn write_discovery_start(&mut self) -> io::Result<()> {
-        self.writeln_message(&format!(r#"{{ "type": "suite", "event": "discovery" }}"#))
+        let mut out = Vec::new();
+        write!(&mut out, r#"{{ "type": "suite", "event": "discovery" }}"#)?;
+        out.push(b'\n');
+        self.writeln_message(&out)
```

Comment:
```suggestion
        self.writeln_message(&format!(r#"{{ "type": "suite", "event": "discovery" }}\n"#).as_bytes())
```

Having to create a `Vec` and then pushing `'\n'` makes sense when there are multiple calls constructing the message, but when it's a single call to `writeln_message` we should be able to keep the `format!` and add `\n` inside of it.

There are other places where the same change should be applied.

Reply:
Why `format!` at all?

Reply:
Fair! I used `format!` when writing the suggestion as the equivalent of `write!`, but indeed it can be removed here (and in other places in this PR).

Reply:
Yes I agree. These things could be improved upon. I've done another pass to clean things up.

Reply:
No, `r#"...\n"#` results in `...\\n`, not a new line at the end of the string.

Reply:
```rust
    let s= r#"{{ "type": "suite", "event": "discovery" }}
"#;
    self.writeln_message(s)
```

Reply:
That'd add a newline correctly, but now the code isn't indented correctly anymore.